Introduce flag to let consul ignore bad check files instead of not starting

varunturlapati opened the pull request:
* Introduce flag to let consul ignore bad check files instead of not starting
* Skipping all bad services or files that would otherwise cause consul agent to fail
* Cleaning up logs and added comments

https://github.com/hashicorp/consul/issues/6444

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul?pullRequest=6445)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul?pullRequest=6445)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul?pullRequest=6445).</sub>

s-christoff commented on agent/agent.go:
```diff
@@ -2854,14 +2854,28 @@ func (a *Agent) deletePid() error {
+		if err != nil && a.config.IgnoreBadCheckFiles == false {
 			return fmt.Errorf("Failed to validate checks for service %q: %v", service.Name, err)
+		} else {
+			if err == nil {
```
I'm a little confused on the approach here. Could you elaborate why we are explicitly doing "else, if err = nil"? My understanding is that instead of an else/if else we could just run the code as normally?

varunturlapati replied:
Hi @s-christoff - for all of the comments the common comment from my side is that I tried to maintain defaults the way consul has. So the new flag works if set to true explicitly. That way it doesn't break compatibility with existing deployments.

Given that inside `if err != nil && a.config.IgnoreBadCheckFiles == false` we're returning, I could skip its corresponding else but I have to have if-else inside, because in one case I append to services adn in the other I simply log that I am skipping that service.

Now, I could invert it to check for
```
if err != nil {
   a.logger.Printf("[WARN] agent: Going to skip service %s ", service)
  } else {
   trimmedServices = append(trimmedServices, service)
}```

I think I started with the sequence of the defaults and deviations from the defaults in my mind and so it looks that way.

Would that address this comment?

s-christoff replied:
That would!

s-christoff commented on agent/agent.go:
```diff
@@ -2872,7 +2886,12 @@ func (a *Agent) loadServices(conf *config.RuntimeConfig) error {
 
 		if err := a.addServiceLocked(ns, chkTypes, false, service.Token, false, ConfigSourceLocal); err != nil {
-			return fmt.Errorf("Failed to register service %q: %v", service.Name, err)
+			if a.config.IgnoreBadCheckFiles == false {
```
Weird nitpick, instead of checking if the flag is set to false, is there anyway we could rework the code to do something if it is true?

varunturlapati replied:
Yeah, I see that here too. My mind working the defaults and deviations from defaults... I can refactor this.

s-christoff replied:
That would be wonderful, thank you!

s-christoff commented on agent/config/builder.go:
```diff
@@ -828,6 +828,7 @@ func (b *Builder) Build() (rt RuntimeConfig, err error) {
 		EncryptVerifyOutgoing:            b.boolVal(c.EncryptVerifyOutgoing),
 		GRPCPort:                         grpcPort,
 		GRPCAddrs:                        grpcAddrs,
+		IgnoreBadCheckFiles:              b.boolVal(c.IgnoreBadCheckFiles),
```
I like this name as a variable, but was thinking perhaps we could incorporate some more clarity by having something like "InvalidCheckFiles". That way if your check files are invalid, you set it to true! What do you think about this?

varunturlapati replied:
I'm not sure I understand this one. My idea here is for SRE/DevOps/Infra folks to set this up in their Consul config so that there is a safeguard against bad checks causing Consul to not start.

So when reading this config param, I feel it clearly indicates that Consul should ignore bad/invalid check files (when true). When false, it doesn't ignore them and so Consul would behave in a default manner. So are you saying I should rename the param?

If not ☝️ , are you suggesting just the attr name to be changed? In that case, I was thinking that reading this variable/attr name in the code would still make it clear to the reader/maintainer, what the var is intending.

So having something like `InvalidCheckFiles` be set to true may only indicate that there are invalid files but not necessarily what should be done when they're encountered.

Does the distinction make sense?
Maybe I didn't understand you correctly. Please clarify.

s-christoff replied:
So, to be *completely* honest I think what I am suggesting the attr name be changed, but I'm not completely against the entire attr name I just dislike the usage of `Bad`.  It seems a tad vague and out of place. `OmitInvalidCheckFailes` or even `IgnoreInvalidCheckFiles` is fine with me, they're both _longer_ but I think a tad more clear.  
Honestly, if it's too much work to change the name, it's no big deal just ignore this comment.

varunturlapati replied:
Oh definitely. (It wasn't clear to me from the original comment.) I think `Invalid` instead of `Bad` is better. At the time of the PR I wasn't sure if we had a formal definition for `invalid`. That said, I should be able to churn out a change real quick.

varunturlapati commented:
@s-christoff I am unable to run tests on master. I wanted to pull the latest master, look at the tests and then rebase my branch and add the tests you recommended.

I filed this issue for now. If you already know what I need to fix, please advise. https://github.com/hashicorp/consul/issues/6762

varunturlapati commented:
Hi @s-christoff - I was able to add the tests and also rebase my code with the latest master.

NOTE:
- I've had difficulty getting all tests to pass on my machine even on the latest master
- What I tried then was to run the tests you asked me to add mine into (on master), confirm that they passed
- On my branch, I then added my tests and tinkered with before-after, and got my new tests to pass too
- Then I rebased with master, squashed my commits and pushed for a PR
- Looks like 1 test is failing and it doesn't look related to my code change.

Could you please take a look and let me know what I could do?

s-christoff commented:
Howdy,
I reran the tests, I'll keep an eye on them and if it doesn't go green we'll dig further into what could be causing it to change.  
_Also apologies for the delay I caught the flu 😷 _

varunturlapati commented:
> Howdy,
> I reran the tests, I'll keep an eye on them and if it doesn't go green we'll dig further into what could be causing it to change.
> _Also apologies for the delay I caught the flu 😷 _

Oh no! Hope you have recovered or at least fairly close to complete recovery.

s-christoff commented:
It looks like you may need to rebase to get this final test to pass - would you be okay with doing that next time you're about to make a push?

varunturlapati commented:
Howdy @s-christoff! Here is the latest update:

1. Renamed the attr/flag to IgnoreInvalidCheckFiles
2. Reran specific tests to get them to Pass
3. Pulled and rebased with latest master
4. Squashed commits and pushed to update the PR

{fingers crossed on the tests}

varunturlapati commented:
Hi @rboyer - thank you for your review! I think I understand some parts of what you said, although I am not sure of the design idea behind why it is that way, for ex.

I think either working on this PR to get it to the shape we want is fine or closing this and using elements of this that are useful in the approach you were hinting at would be fine. Either case, I spoke to our TAM (@cstanlee - Carlos Stanley, in case I @-refed the wrong person). He agreed to setup a meeting for us to go over how to proceed.

jsosulska commented:
Hello @varunturlapati varunturlapati,

Thanks for all the hard work on this PR. Ultimately, this solution is not one that is going to be implemented. [The original ticket](https://github.com/hashicorp/consul/issues/6444) wants to mitigate the chance of Consul being unable to start because of a malformed or incorrect service registration file. The easiest way to do this is to not use the the configuration directory as a way of registering services. Instead, as part of the deployment, script/use a CI tool to POST to the service definition to `agent/service/register` endpoint of the local Consul agent to register the service. If the service you are deploying doesn't start after reboot, the service's script check will let you know when Consul comes back up. If the API rejects the service definition, the local Consul agent is unaffected.

While performing a service registration via a config file is a valid method of registration, it is less than ideal. Think of the config-dir as "Agent Configuration".   Being able to write to disk and trigger a reload requires some fairly high privileges. By gating registration through other methods, like a POST method with an ACL token, there is many more safeguards in place.

A more conservative improvement can be found in https://github.com/hashicorp/consul/issues/7560. This would allow a `consul validate` * to run a validation on the service registration files from the Consul binary. A script/CI step to run the consul validate command on the service registration before deployment, or before restart/reload, would achieve the same desired result in the issue ticket much more quickly.

*Currently, consul validate only works on agent configuration.
In the meantime, we'll close out #6444 and #6445. 
Thank you again for the effort you put into this work 👏 
Happy Coding,
Jono
